### PR TITLE
Make bitvec helpers available as global, don't polute top-level namespace

### PR DESCRIPTION
**Pull request by vlstill**
- remove code duplication
- `count_{leading,trailing}_zeroes` + `popcount` are now available globally to use on built-in integral types
- builtin_* helpers + the aforementioned new functions are defined in `namespace bv` to avoid possible collisions

### CODE REVIEW

File: lib/bitvec.h [hunk 1]
```diff
@@ -55,6 +57,48 @@ static inline int builtin_popcount(unsigned long x) { return __builtin_popcountl
 static inline int builtin_popcount(unsigned long long x) { return __builtin_popcountll(x); }
 #endif
 
+template<typename I>
+int count_trailing_zeroes(I x) {
```
Comment (davidbolvansky) [inline lib/bitvec.h:61]: Any protection against zero?

File: lib/bitvec.h [hunk 2]
```diff
@@ -55,6 +57,48 @@ static inline int builtin_popcount(unsigned long x) { return __builtin_popcountl
...
+    static_assert(std::is_integral<I>::value, "'I' has to be integral type");
```
[the remark this answers is not in the document]
Comment (vlstill) [reply lib/bitvec.h:62]: Added an `assert`. I did not want to add BUG_CHECK for such a low-level functionality & there are already asserts in bitvec.h.

File: lib/bitvec.h (hunk 1 above)
Comment (vlstill) [reply lib/bitvec.h:61]: Added an `assert`. I did not want to add BUG_CHECK for such a low-level functionality & there are already asserts in bitvec.h.

**Comment by vlstill** [comment]
@davidbolvansky, it appears I have accidentally removed your comment asking why I have used template. Sorry about that, I wanted to delete my misplaced comment and apparently remove the thread (why am I even allowed to do this I don't understand).

The reply: Template to avoid duplicated code... I could probably restrict it more (i.e. unsigned & <= 64 bits) to ensure the proper builtin wrappers should be available. To be even more safe we could do something like `template<typename I> builtin_ctz(I) = delete` to avoid any implicit casts (as *it appears* that things like `uint32_t` or `uintptr_t` have to be typedefs so it should not produce false errors).

File: lib/bitvec.h (hunk 1 above)
Comment (davidbolvansky) [reply lib/bitvec.h:61]: Thanks

**Comment by vlstill** [comment]
@davidbolvansky, I did make the static assert a bit more comprehensive.

**Review by mbudiu-vmw — APPROVED** [review]

**Pull request merged**